### DISCUSSION TITLE
New Rule: Find AND Select All Matches

### DISCUSSION THREAD

**Discussion by FelipeRearden** [jglev/obsidian-apply-patterns-plugin#9] [Ideas]
[More info about me here](https://github.com/publicus/obsidian-apply-patterns-plugin/discussions/3)

Hello @publicus!!!

In some cases, we don’t need to replace the find matches, we need to do something with all the matches found by the `find` feature: `cut`, `copy` to transfer all of the matches. 
  
My request is a new rule to `find` and `copy/cut all matches`.

### Showing my workflow in CotEditor:
This is a chunk of my raw text:
![Screen Shot 2021-08-10 at 09 18 02](https://user-images.githubusercontent.com/85180696/128865440-0196e001-0a26-4934-8124-80c710a8e6b8.png)

#### My goal is to: 
1- find every DEVONthink link `[L4](x-devonthink-item://C5A952D4-8CB9-4CA9-9B3A-BC397B08C1F6?line=3)`
2- Select All of them
3- Copy to clipboard
4- Delete only `(x-devonthink-item://C5A952D4-8CB9-4CA9-9B3A-BC397B08C1F6?line=3)`
5- Paste `[L4](x-devonthink-item://C5A952D4-8CB9-4CA9-9B3A-BC397B08C1F6?line=3)` in the end of the note
6- Select All + Copy + Paste in Obsidian

#### In CotEditor 
A- find every DEVONthink link `\[L\d{1,3}\]\(x-devonthink-item:\/\/\w{8}\-\w{4}\-\w{4}\-\w{4}\-\w{12}\?line=\d{1,2}\)`
![Screen Shot 2021-08-10 at 09 10 42](https://user-images.githubusercontent.com/85180696/128865517-a8924a55-d6cf-4c4b-847e-ee5e7f3b3574.png)

B- Select All of them
![Screen Shot 2021-08-10 at 09 11 10](https://user-images.githubusercontent.com/85180696/128865533-f2215f72-c2e1-4be7-b78a-1903914a802c.png)

C- Copy to clipboard `cmd+c`

D- Delete only `\(x-devonthink-item:\/\/\w{8}\-\w{4}\-\w{4}\-\w{4}\-\w{12}\?line=\d{1,2}\)` [find and replace with ""]
![Screen Shot 2021-08-10 at 09 14 51](https://user-images.githubusercontent.com/85180696/128865562-3edcd2b6-53fc-4a89-a3b4-6a9ac0aa761f.png)

E- Paste clipboard in the end of the note
![Screen Shot 2021-08-10 at 09 16 13](https://user-images.githubusercontent.com/85180696/128865590-1df4097a-928b-4630-b685-f50b6d18a23c.png)

F- Select All and Copy/Paste in Obsidian

### Feature request: New Rules

if we have a new rule to `find` and `Copy All Matches` / `Cut All Matches` I could replicate this workflow in Obsidian:

Steps A + B + C -> Rule 1
Step D -> Rule 2
Step E could be achieved manually 
Step F is not necessary anymore 

Let me know if you need more information.

Thanks and have a great day!

**Comment by jglev** [comment]
Thank you for writing, @FelipeRearden!

Utilizing the clipboard is something I've been thinking about, too. I think that there's a powerful feature to be found in that idea, but every time I've started drafting ideas around it over the past couple of weeks, it seems to me like new questions about interface and behavior present themselves.

The workflow you describe is specific enough that it strikes me that it may warrant its own plugin. Since currently, Obsidian's `Search: Search in all files` command supports both regular expressions and copying results, I'm _currently_ planning to hold off on implementing clipboard support in this plugin until I have a better sense of additional problems it could address. At the least, though, for the workflow you described, you could do the following currently, I think:

> 1- find every DEVONthink link [L4](x-devonthink-item://C5A952D4-8CB9-4CA9-9B3A-BC397B08C1F6?line=3)

Use `Search: Search in all files`, writing `/your regex from above/ file:NameOfFile`

> 2- Select All of them
> 3- Copy to clipboard

Use the "Copy search results" button within the search sidebar.

> 4- Delete only (x-devonthink-item://C5A952D4-8CB9-4CA9-9B3A-BC397B08C1F6?line=3)

Use this plugin (Apply Patterns) to replace matches with nothing.

> 5- Paste [L4](x-devonthink-item://C5A952D4-8CB9-4CA9-9B3A-BC397B08C1F6?line=3) in the end of the note
> 6- Select All + Copy + Paste in Obsidian

Perform both manually, as you also suggested above.

Does that process look accurate to you, too?

Again, thank you for writing!

**Comment by FelipeRearden** [comment]
Hello again @publicus  !!!!

Fist of all, thank you for you reply. If don't mind I'm gonna reply your replies and make a "new" feature request explaining to you my real need :)

I wanna make sure that we are on the same page about this :)

> Utilizing the clipboard is something I've been thinking about, too. 

Let's forget about `copy to cilpboard` and replace this to `select all matches` in `find/from` (I will explain in the end). Maybe what I am suggesting is somenthing feasible :)

> At the least, though, for the workflow you described, you could do the following currently, I think:

Thanks for helping with a workaround!!!! I spend so much time with plug-in that I forgot about this. Two things about the Search Plug-in:
- Unfortunately `copy search results` don't copy the search results. It copies the file/filepath of the search results :) (I think it's a bug). Take a look:
![Screen Shot 2021-08-12 at 10 01 44](https://user-images.githubusercontent.com/85180696/129209856-08594dd4-ca24-4dbd-a37d-10be4d868b38.png)
![Screen Shot 2021-08-12 at 10 02 23](https://user-images.githubusercontent.com/85180696/129209906-b9c04824-00b6-4fb4-b3a6-1ce5400b4ae3.png)
- `copy search results` don't have a Command Palette `command`. This way I can't set a hotkey for this (more about this in the end)

### A New Request
If we could have a new rule or a new checkbox inside a rule that:
- `Find/from` + `Select All Matches` found in `Find/from`
- No `replace/to` (it "just" find and select all matches) just like this image
![Screen Shot 2021-08-10 at 09 11 10](https://user-images.githubusercontent.com/85180696/129215717-5dbd1aa8-0e40-44bc-8704-580a8b902396.png)

### Why I'm asking for All of this new Features

My main ideia is to apply this workflow using only hotkeys. 

There is a new plug-in called [Quickadd](https://github.com/chhoumann/quickadd/blob/master/docs/Choices/MacroChoice.md) that has a fantastic feature: the user can create macros to stack Obsidian Commands and run then with a hotkey. One hotkey to run 5 commands for example.
![Screen Shot 2021-08-12 at 10 16 47](https://user-images.githubusercontent.com/85180696/129210282-0bdacc16-cab6-4ed9-8e1a-1425a6bea7cc.png)

Hotkey for the MACRO:
![Screen Shot 2021-08-12 at 11 03 00](https://user-images.githubusercontent.com/85180696/129210880-4d4d1fc7-6d60-4189-bb71-7515b2f3c91a.png)

This way I can create a MACRO to transfer text from DEVONthink to Obsidian with one or two hotkeys. Look what I have today for creating a new note in Obsidian with a specfic Template 
- Open my workspace for writing/reading web articles;
- Create a new note with a template for web articles in a specific folder;
- `Paste` the content from CotEditor copied to the clipboard **(that's where `Apply Patterns` hotkeys will save me** :));
- Rearrange  cursor position and split the note into two windows.

After paste the raw text from DEVONthink, I will use the new `Apply Patterns` hotkeys and `whole document` to clean my raw text and forget about CotEditor.  

Here is the magical thing: QuickAdd has 5 exclusive commands that are not in the Command Pallete `Copy` `Paste` `Cut` `Select Active Line` `Select Link an Active Line`
![Screen Shot 2021-08-12 at 10 18 20](https://user-images.githubusercontent.com/85180696/129211032-f932ecb4-40b8-4bb3-9b16-da93476d987c.png)

This way, with a new feature on `Apply Patterns` that Select all Matches from `find/to` I could use `Copy` from QuickAdd to store these matches and do step E (Paste clipboard in the end of the note) manually 
- There is a plug-in called [Jump to Link](https://github.com/mrjackphil/obsidian-jump-to-link#how-to-use-jump-to-anywhere-editor-navigation) that could help me to jump to a place in the end of the note and `Paste` the Selected Matches (I didn't tried yet)
- Maybe Step E could be automate

That's why I said before about forgeting about `copy to clipboard` and this is why your plug-in is so important to me. :)

### Sorry to bother you with my "madness"

Please I don't wanna make you feel uncomfortable about this. I really just wanna share with you one of my usage of you plug-in. 

I hope this is feasible but feel free to decline this feature request. I'm glad to share with my workflow and I am very very happy for the new features in 1.2.0.

Sorry for spending your time with this and sorry for my english mistakes (I'm not a native speaker but I tried to write as right as I know)

Thanks for everything @publicus !!!!

**Reply by jglev** [reply]
@FelipeRearden I haven't forgotten about this! Thank you for taking the time to write up your idea in such detail. And I don't feel uncomfortable; thank you for your sensitivity! : )

I'm still thinking about this -- I agree that having the ability to find and select all matches could be useful, but I remain unsure whether it falls within the scope of this plugin (vs. making another plugin to add just that functionality). It may fall in-scope! I just have still been mulling on it.

**Reply by FelipeRearden** [reply]
Hello @publicus !!!!

Thank you very much for your reply! I’m glad to know that I’m not bothering you :) 
It brings me a good feeling :)

Thanks for considering this feature. I have been thinking about this too and how this fits into to this amazing plug-in: 

### My Perspective  

Apply Patterns is a strong and robust plug-in that do everything right. It’s a world class plugin :)

**RegEx with all option that is possible** 
- global
- Multiline
- Sticky 
- Insensitive 
- ‘/n`

**Automation and freedom with the new `Commands` for hotkeys**

**User friendly interface to create/modify and understand each of one of the options**

**Info button to evaluate the expressions (this is amazing !!!!!!)**

**No one is leaving behind:**
- whole document
- Selection 
- Active line 

**My humble opinion is that `find` and `select all matches` is the final piece that closes the core features** 
- User can `find` and `replace` 
- User can `find` and `delete` (replace with “”) 
- User can `find`and `select` these results to do many things as (copy/cut to use in other document, see how many times the pattern occurs in the document)

There is nothing the  user might need. Everything is observed :)`find/replace/delete/select.`

—-

Let me know if you wanna share some thoughts. We can talk on the Obsidian Discord Server too :) 

Thanks again for your hard work with this plug-in and for being such a great person :)

Thanks for giving me the opportunity to share this with you. With this feature or not I’m glad to have this amazing plug-in in my workflow everyday :)

Have a great Sunday !

**Reply by jglev** [reply]
@FelipeRearden, thank you for writing up your thoughts! For what it's worth to know, your argument convinced me : ) I'm still thinking through how I'd like to implement it (I'm still thinking through whether using the clipboard might make this a more generalized, useful feature), but I'm on board.

**Reply by FelipeRearden** [reply]
@publicus  , this is amazing, I’m so happy !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

I’m 100% sure that you will find the best solution for the user. it’s clear to me that every detail of Apply Patterns was built with attention and care. :)

Thanks for your patience with my FR. Let me know if I can help. I’m always here :) 

Thanks again and have a great day !
